cmd/kubeadm/app/cmd/upgrade/plan.go

```diff
@@ -97,131 +99,154 @@ func runPlan(flags *planFlags, userVersion string) error {
+	components = append(components, newComponentUpgradePlan(constants.KubeControllerManager, up.Before.KubeVersion, up.After.KubeVersion))
+	components = append(components, newComponentUpgradePlan(constants.KubeScheduler, up.Before.KubeVersion, up.After.KubeVersion))
+	components = append(components, newComponentUpgradePlan(constants.KubeProxy, up.Before.KubeVersion, up.After.KubeVersion))
 
-		switch upgrade.After.DNSType {
-		case kubeadmapi.CoreDNS:
-			printCoreDNS = true
-			coreDNSAfterVersion = upgrade.After.DNSVersion
-		case kubeadmapi.KubeDNS:
-			printKubeDNS = true
-			kubeDNSAfterVersion = upgrade.After.DNSVersion
-		}
+	components = appendDNSComponent(components, up, kubeadmapi.CoreDNS, constants.CoreDNS)
+	components = appendDNSComponent(components, up, kubeadmapi.KubeDNS, constants.KubeDNS)
 
-		if printCoreDNS {
-			fmt.Fprintf(tabw, "CoreDNS\t%s\t%s\n", coreDNSBeforeVersion, coreDNSAfterVersion)
-		}
-		if printKubeDNS {
-			fmt.Fprintf(tabw, "Kube DNS\t%s\t%s\n", kubeDNSBeforeVersion, kubeDNSAfterVersion)
-		}
+	if !isExternalEtcd {
+		components = append(components, newComponentUpgradePlan(constants.Etcd, up.Before.EtcdVersion, up.After.EtcdVersion))
+	}
 
-		if !isExternalEtcd {
-			fmt.Fprintf(tabw, "Etcd\t%s\t%s\n", upgrade.Before.EtcdVersion, upgrade.After.EtcdVersion)
-		}
+	return &outputapiv1alpha1.UpgradePlan{Components: components}, unstableVersionFlag, nil
+}
+
+// printUpgradePlan prints a UX-friendly overview of what versions are available to upgrade to
+func printUpgradePlan(up *upgrade.Upgrade, plan *outputapiv1alpha1.UpgradePlan, unstableVersionFlag string, isExternalEtcd bool, w io.Writer) {
+	// The tab writer writes to the "real" writer w
+	tabw := tabwriter.NewWriter(w, 10, 4, 3, ' ', 0)
 
-		// The tabwriter should be flushed at this stage as we have now put in all the required content for this time. This is required for the tabs' size to be correct.
+	// eOT (end Of Table) helper
+	eOT := func() {
 		tabw.Flush()
 		fmt.Fprintln(w, "")
-		fmt.Fprintln(w, "You can now apply the upgrade by executing the following command:")
-		fmt.Fprintln(w, "")
-		fmt.Fprintf(w, "\tkubeadm upgrade apply %s%s\n", upgrade.After.KubeVersion, UnstableVersionFlag)
-		fmt.Fprintln(w, "")
+	}
 
-		if upgrade.Before.KubeadmVersion != upgrade.After.KubeadmVersion {
-			fmt.Fprintf(w, "Note: Before you can perform this upgrade, you have to update kubeadm to %s.\n", upgrade.After.KubeadmVersion)
-			fmt.Fprintln(w, "")
+	printHeader := true
+	printManualUpgradeHeader := true
+	for _, component := range plan.Components {
+		if isExternalEtcd && component.Name == constants.Etcd {
+			fmt.Fprintln(w, "External components that should be upgraded manually before you upgrade the control plane with 'kubeadm upgrade apply':")
+			fmt.Fprintln(tabw, "COMPONENT\tCURRENT\tAVAILABLE")
+			fmt.Fprintf(tabw, "%s\t%s\t%s\n", component.Name, component.CurrentVersion, component.NewVersion)
```

Comment:
It's best to keep the user friendly component names if we can. For that matter, we can perform a quick translation here by using a static `map[string]string` in a helper func. Something like so:
```golang
var userFriendlyComponentNamesMap = map[string]string {
	constants.KubeAPIServer: "API Server",
	// ...
}

func getUserFriendlyComponentName(component string) string {
	if name, ok := userFriendlyComponentNamesMap[component]; ok {
		return name
	}
	return component
}

// ...
fmt.Fprintf(tabw, "%s\t%s\t%s\n", getUserFriendlyComponentName(component.Name), component.CurrentVersion, component.NewVersion)
// ...
```

Reply:
AFAIR this was discussed in one of the previous PRs of this kind and we decided to go with the names from constants.go. @neolit123 was ok with this.

Reply:
i don't recall, my personal preference is to have everything coming from constants - e.g. `kube-apiserver, kubelet...`. our `upgrade apply` output is doing that.

Reply:
ok, thanks. Let's agree not to have different names for the same component then. @rosti would it be ok for you?

Reply:
It's possible that some existing tools actually parse those tables at the moment. Hence, changing `API Server` to `kube-apiserver` might prove problematic for some users. That's my only concern right now. Otherwise, I am fine with using the consts in the user readable output.

@detiber by any chance, is Cluster API doing any parsing of this table (`kubeadm upgrade plan` that is)?

Reply:
We are not currently doing any parsing of this output as we currently use a loose coupling with `kubeadm`.

Reply:
@rosti re map or not to map names, I'm ok with either approach. Just let me know what you decided.

Reply:
@rosti @neolit123 as for introducing groups of components to the UpgradePlan structure. I'm proposing either not to do it at all (see my reasons explained above) or at least name it differently.

Reply:
@bart0sh I am fine with not remapping.
Also, due to kubernetes/kubeadm#2058 external etcd will likely not be reported anymore by `kubeadm upgrade plan`. That leaves only the kubelet as of today as a component that needs manual upgrading.
Hence, let's merge it as is. If we need to add differentiation we can do that in alpha2 of the output API.

/approve
/lgtm